First stab at a TS conversion

jods4 opened the pull request:
I decided to try to convert one project to Typescript to see how things work out (build commands, etc.).

It's not complete: I have to add `lint`, `test` and `dev` tasks -- help is appreciated.

Until `tools-typescript` PRs are merged, I have simply copy-pasted the `tsconfig.base.json` in this project. It's meant to be replaced by a reference.

Altough this is a very small and simple project, there is much to be said about the actual TS code. For now I just did some minimal changes so that it compiles without errors. Please ignore that, I'll open a separate discussion about the code, just focus on the process here.

On the upside, there are a lot less dependencies, a lot less files, build is much simpler. Also it's immediately apparent that getting strict type checking and good definitions is going to be helpful.

**Please have a good look at the changes, because some are non-trivial (I'll let you discover them). As this is our first conversion, reviews and feedbacks are much appreciated.**

There are a few things that I don't know how to change or if they're ok, like the jspm and bower parts, as well as CircleCI.

jods4 committed:
First stab at a TS conversion

bigopon commented on src/index.ts:
```diff
@@ -362,8 +362,8 @@ export function initializePAL(callback: (platform: Platform, feature: Feature, d
   isInitialized = true;
-  if (typeof Object.getPropertyDescriptor !== 'function') {
-    Object.getPropertyDescriptor = function(subject, name) {
+  if (typeof (Object as any).getPropertyDescriptor !== 'function') {
```
Shouldn't we extend the interface of Object here ? Or is this because we just want it to be used internally?

jods4 replied:
If it's meant to be public, yes absolutely.
And you're spot on, this particular line is something I'd like to discuss.

I'll open a PR with code changes at a later point. I wanted to get the "infrastructure" of the migration first.

EisenbergEffect commented:
@jods4 We had a standard setup we were using on all the TS projects. Did you use that?

StrahilKazlachev commented:
@jods4 Did you commit the *dist* intentionally?

jods4 commented:
@EisenbergEffect No, I wasn't aware of that! I did not know we had a few TS projects already.
I took the existing JS/gulp project and transformed it into TS/npm.
Can you point me to the standard setup? I'll compare with what I've done here.

@StrahilKazlachev This time, yes. To show and compare the difference in output.
For some reason, the JS build produced an extra `index.js` that simply re-exported a `aurelia-pal.js` file.
Not sure why as the `main` entrypoint in `package.json` actually was `aurelia-pal.js` not `index.js`.

The TS build only outputs a single `index.js` and I set it as `main` in `package.json`.

This is for the best but like _any_ change it is a potential breaking change if someone used exact file paths in their build or something like that.

StrahilKazlachev commented:
@jods4 It's not still a *standard* setup, but check the `aurelia-valdiation` and `aurelia-dialog` - the `package.json` has a load of npm scripts. In the dialog repo I've tweaked a bit the setup(don't remember where) I copied from validation.
It would be great in the end to have all common parts in `aurelia-tools-ts` - configs, utils, guidelines, etc. Why not even a whole template.

jods4 commented:
I'll have a look at `aurelia-dialog` then.

> It would be great in the end to have all common parts in aurelia-tools-ts - configs, utils, guidelines, etc. Why not even a whole template.

Absolutely. While doing that, I was really wondering if there's any way to move the scripts in `npm` out of this repo and into `aurelia-tools-ts` so that they're not duplicated everywhere. Not sure what the best approach is, though. Would be easier with a scripted build (... gulp? damn! 😭 ).

Other random thought: a benefit of mono-repo is that we could share the `.editorconfig` and `.gitignore` at root. Those things can't be linked to like `tsconfig.json` can, they have to be duplicated.

EisenbergEffect commented:
We've been basing most of the new work off of aurelia-validation.

StrahilKazlachev commented:
I would say *most* is an understatement - it's more like *all* 😅.
I was also wondering how can the script part be shared, and gave up 😢.
`tslint.json` can also be linked, and we should do it next. It will be much more work.

jods4 commented:
I had a look at `aurelia-validation` and `aurelia-dialog`. I am _very_ pleased how similar to my own they are -- given we did those independently!

Small differences I noted:
- They have the lint, docs, tests and publishing tasks I haven't done yet...  I'm just going to steal those 😁 
- Cleaning is done for each target instead of all of them in one swoop. Will do the same.
- There's a consistent `cross-env` usage that I just don't get. None of the commands set an env. variable, so what's the point? I'm doing fine without?
- Definition files are produced in each target, which is wasteful, I only produce one. I will move it to a sub folder, though, because I realized that there might be more than one file (a future version of TS should fix that, but in the meantime...)
- Targets are built concurrently, I do them in sequence. Will change that as well. What benefits does `concurrently` have compared to a simple `&`?

EisenbergEffect commented:
@jdanyow would be the best person to engage on this since he made the original setup.

jods4 committed:
Folder for types, clean per target, parallel build.

jods4 committed:
Add linter

jods4 commented:
Added linter command. I copied the lint configuration from `aurelia-dialog` inside `tslint.base.json`. 
This file isn't meant to stay here, we'll either link to it inside `aurelia-tools-ts` or put it at the root of a mono-repo.

Linter complains, of course. I'll fix.

jods4 committed:
Linter is happy now

jods4 commented:
Linter is fixed. 
Had to change a lot, I added 3 exceptions with `tslint:disable-next-line` comments. 
You can check them out, worth thinking if we should tweak our rules. I think `only-arrow-functions` could be relaxed as we compile with `noImplicitThis` which means our code is `this`-safe anyway. Yes I just made it up.

StrahilKazlachev commented:
@jods4 If you are interested you could open a PR in tools-ts, or I'll do when I get time.
I really think we should setup our own tslint.json(rules), don't want to link to the ones in `node_modules/tslint`. But we can use one of those as starting point.

jods4 commented:
@StrahilKazlachev I'd like others' opinion before making PR changes to existing ruleset.

gheoan commented:
Maybe something like [yamscripts](https://www.npmjs.com/package/yamscripts) can be used to share the npm scripts.

What is team's opinion on monorepo and gulp instead of npm scripts?

jdanyow commented:
Re: cross-env and concurrently:

Build needs to work on windows. Don’t want to alienate contributors, makes it easy to run the test suite in edge and IE

jods4 commented:
@jdanyow I am on Windows! 😁 
My understanding is that `cross-env` is useful to set env. variables cross-platform, but we set none.
`concurrently`... I don't know? `&` seems to work fine?

@gheoan Thanks for the tips!
yamscripts generates scripts from a template, so you're saying we could generate from a common template, right?
Monorepo is something we're debating!
If you look at the diff, you'll find gulp was used before npm. But it's hard to justify taking a dep on gulp & friends when your "build" basically comes down to calling `tsc`.

I found this and I have to say it looks interesting:
https://github.com/testdouble/scripty
It allows cleaner scripts, advanced options _and scripts sharing!_

jdanyow commented:
I don't care if the build is concurrent or if we use cross-env/concurrently. Just saying whatever we do for scripts needs to work on windows. I think cross-env is making "&" work today. If there's another way that's better let's use it and backport into validation etc.

jods4 committed:
Publish tools

gheoan commented:
@jods4 Yes, but the tool you found seems like a better solution.

One advantage of `gulp` (at least in version 4) is that tasks are functions so they can be shared. In theory the `ts-tool` package can export `gulp` tasks that can be used by the other packages.

I guess it comes down to depending on gulp plugins vs depending on tools to normalize the npm scripts commands. I think `concurrently` is required so that & can work on powershell on windows.

Thank you for taking the time to experiment!

jods4 committed:
Added npm scripts to run tests

jods4 commented:
Added tests commands.

One thing I disliked was using JSPM for testing because it meant having to install all dependencies yet one more time. So it made me happy to see that `aurelia-dialog` and `aurelia-validation` have moved away from jspm and replaced it with requirejs.

I did try to improve a little more by replacing `karma-requirejs` by `karma-typescript`:
- Little config, no additional entry point file like requirejs.
- No need for extra distributions or whatnot, reads directly from the `src` and `test` typescript files.
- Watches both src and tests for reruns.
- Produces code coverage in the original typescript files (see below).

What do you think?

`karma-typescript` supports many different code coverage tools, here's the default out-of-the-box output:

![image](https://user-images.githubusercontent.com/3832820/32693247-e13fc938-c727-11e7-9a7f-4eaa1f9339da.png)

jods4 committed:
Updated Circle CI tests and README

jods4 committed:
Added docs

jdanyow commented:
Looking good, I like the usage of `extends` in the tsconfig.

* When I execute `npm run test` there's an compile error in the output. The test succeeds:
    ```
    18:38 $ npm run test

    > aurelia-pal@1.4.0 test /home/jd/repos/aurelia/pal
    > karma start --single-run

    12 11 2017 18:38:57.117:INFO [compiler.karma-typescript]: Compiling project using Typescript 2.6.1
    12 11 2017 18:38:59.390:ERROR [compiler.karma-typescript]: test/dom.spec.ts(1,26): error TS2307: Cannot find module 'aurelia-pal'.

    12 11 2017 18:38:59.437:INFO [compiler.karma-typescript]: Compiled 3 files in 2300 ms.
    12 11 2017 18:39:00.049:INFO [bundler.karma-typescript]: Bundled imports for 2 file(s) in 111 ms.
    12 11 2017 18:39:00.129:INFO [karma]: Karma v1.7.1 server started at http://0.0.0.0:9876/
    12 11 2017 18:39:00.130:INFO [launcher]: Launching browser Chrome with unlimited concurrency
    12 11 2017 18:39:00.136:INFO [launcher]: Starting browser Chrome
    12 11 2017 18:39:01.366:INFO [Chrome 62.0.3202 (Linux 0.0.0)]: Connected on socket h1kG71mKL4Ve291HAAAA with id 8129961
    Chrome 62.0.3202 (Linux 0.0.0): Executed 1 of 1 SUCCESS (0.003 secs / 0.003 secs)
    ```

* Could update typedoc to 0.9.0 while you're in there.

* re: karma-typescript setup for the tests- looks good to me for a repo like this. I'd like to see if this can be applied to a repo like aurelia-validation that needs the loader, etc so it can run component tests using aurelia-testing

* does this work in powershell and cmd? Looks like you might be using git bash?

jods4 commented:
@jdanyow One by one:

Test compilation fails: at first I was baffled because it works fine here. But I think I have a guess... I have the `aurelia-pal.d.ts` file somewhere higher up in a shared `node_modules`. My guess is that you don't (which is normal) and so TS compiler doesn't have type info for the import. I will try to fix... not sure how.

Upgrade Typedoc: I tried and it was a mess. I talked with @StrahilKazlachev and he had the same experience and stopped trying. Unless someone wants to spend time on this, we're stuck at 0.8.0.

Tests with karma-typescript: I did some tests and talked with @zewa666. It doesn't work well if you try to run the full Aurelia framework (e.g. use `aurelia-testing`). If you do, you basically have the usual `aurelia-loader` situation: you have to ensure it can load every dependency, which requires lots of config and appropriate libs, etc. We can discuss this more thoroughly in gitter, but my feeling is: let's go with `karma-typescript` for the vast majority (if not all) of the core libs that are pure tests without `aurelia-loader`. There are lots of benefits here: simpler, leaner, good code coverage support. For repos that need `aurelia-loader` such as `validation` or `dialog`, we'll have to switch gears and use something more complex (the current setup is with `karma-requirejs`.

Windows: I've done all of this in `cmd` and it worked fine. For the sake of it I just tried a build in Powershell and it worked as well.

jods4 committed:
Fix in docs scripts by @StrahilKazlachev

jods4 committed:
Add tslint language service for IDE support.

zewa666 commented:
I know this is getting quite big, but as we discussed recently, how about adding husky + a prepush/precommit hook (whichever you think suites better) which calls tslint. Ideally the prepush hook would additionally run the tests to make sure that no *unstable* version gets pushed to github

mjwwit commented:
@zewa666 or you could add CircleCI to the PR checks on GitHub and run the linter there. It'll save you and every contributor a bit of tooling and config. IMO we should make it as easy as possible to create a PR, there's enough hurdles to go through already.

jods4 commented:
Given the choice between blocking commit and a failing check in the PR that prevents merging, I would personnally lean toward the 2nd one. 
Sometimes you want to commit stuff for discussion even if it's not right, this PR is a good example of that.

Once this is decided I will happily merge CircleCI config or Husky hooks but I need a little help/contribution as I know nada about both of those.

mjwwit commented:
@jods4 I can't seem to get your karma config working for projects with templates. The change to `karma-typescript` instead of `karma-requirejs` may be the issue.
I'll continue to work on it, but I may need some help. I'm trying this on the aurelia-testing project.

jods4 commented:
@mjwwit See my comment above https://github.com/aurelia/pal/pull/27#issuecomment-344001029

`aurelia-loader` doesn't play nice with the `karma-typescript` loader and after quite some time spent on this @zewa666 and I have given up on making them work together.

Most tests of Aurelia core are pure JS unit tests, so we suggest that we use `karma-requirejs` for them: it's simpler and has good built-in code coverage.
If your tests need to use `aurelia-loader` a more complex setup is required, `aurelia-validation` and `aurelia-dialog` are based on `karma-requirejs` and they work, so that's the baseline currently.

mjwwit commented:
@jods, Ah, I didn't see your update.

I still had to update the karma config used in `aurelia-validation` to be able to load templates.  The result can be seen in https://github.com/aurelia/testing/blob/typescript-port/karma.conf.js.
Are we aiming for a common config that works for all projects?

jods4 commented:
Common config -> yes, for sure. The goal is to have the same organization everywhere.

Now I personnally think that if there are compelling advantages we could go for 1-2 standard variations, for example one simple test setup with little config, no intermediate files, live watching of both sources and tests files and good code coverage reports; and one setup supporting aurelia-loader.

What do you guys think?

StrahilKazlachev commented:
That makes sense since the majority of the projects are *code only*.
We should also check why some projects merge their files when generating the *dist*, is it still needed(one step less)?

zewa666 commented:
Fully agree, two test solutions are ok until we find a better way

EisenbergEffect commented:
What's the status on this?

jods4 commented:
@EisenbergEffect I think we reached a point that's not too bad to start with. I'm sure there will be changes down the line, but that's how it works.

**This PR is not meant to be merged.** It was meant for discussion and setting up the new project structure, I think it did its part.

Next steps probably are:
1. set up the mono-repo structure we want to adopt;
2. "commit" that PR work over there.
3. Finalize the TS port of `aurelia/pal`. I worked on the project structure mostly. Code compiles without error but surely could be made a little nicer.

EisenbergEffect commented:
Sounds good. I was thinking monorepo was the next step too so I’ll help move that ahead soon.

Sent from my iPhone

> On Dec 11, 2017, at 10:54 AM, jods <notifications@github.com> wrote:
> 
> @EisenbergEffect I think we reached a point that's not too bad to start with. I'm sure there will be changes down the line, but that's how it works.
> 
> This PR is not meant to be merged. It was meant for discussion and setting up the new project structure, I think it did its part.
> 
> Next steps probably are:
> 
> set up the mono-repo structure we want to adopt;
> "commit" that PR work over there.
> Finalize the TS port of aurelia/pal. I worked on the project structure mostly. Code compiles without error but surely could be made a little nicer.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

EisenbergEffect commented:
Closing this out since we'll prefer to focus on vNext.
